File: year2019/go/pkg/coders/channelCoder.go
package coders

import (
	"context"
	"fmt"
	"math"
)

type observer interface {
	notify(output int)
}

type channelCoder struct {
	reader     chan int
	codes      []int
	idx        int
	identifier string
	output     []int
	observer   observer
}

func newChannelCoder(identifier string, codesInput []int, inputs []int) channelCoder {
	codes := make([]int, len(codesInput))
	copy(codes, codesInput)
	reader := make(chan int, 10+len(inputs))
	for _, v := range inputs {
		reader <- v
	}
	return channelCoder{
		reader:     reader,
		codes:      codes,
		idx:        0,
		identifier: identifier,
	}
}

// Notify implements observer
func (cc *channelCoder) notify(output int) {
	go func() {
		cc.reader <- output
	}()
}

func (cc *channelCoder) getLatestSignal() int {
	return cc.output[len(cc.output)-1]
}

func (cc *channelCoder) attach(ctx context.Context, obs observer) <-chan struct{} {
	cc.observer = obs
	done := make(chan struct{})
	go cc.run(ctx, done)
	return done
}

func (cc *channelCoder) run(ctx context.Context, done chan<- struct{}) {
	for {
		select {
		case input, ok := <-cc.reader:
			if !ok {
				cc.reader = nil // turn this case when channel closed
				continue
			}
			inputUsed := false
		optLoop:
			for {
				execution := cc.codes[cc.idx]
				switch optCode := execution % 100; optCode {
				case 1:
					cc.codes[cc.getIdxFromMode(execution, 3, cc.idx)] =
						cc.codes[cc.getIdxFromMode(execution, 2, cc.idx)] + cc.codes[cc.getIdxFromMode(execution, 1, cc.idx)]
					cc.idx += 4
				case 2:
					cc.codes[cc.getIdxFromMode(execution, 3, cc.idx)] =
						cc.codes[cc.getIdxFromMode(execution, 2, cc.idx)] * cc.codes[cc.getIdxFromMode(execution, 1, cc.idx)]
					cc.idx += 4
				case 3:
					if inputUsed {
						break optLoop
					}
					cc.codes[cc.getIdxFromMode(execution, 1, cc.idx)] = input
					cc.idx += 2
					inputUsed = true
				case 4:
					output := cc.codes[cc.getIdxFromMode(execution, 1, cc.idx)]
					cc.observer.notify(output)
					cc.output = append(cc.output, output)
					cc.idx += 2
				case 5:
					if cc.codes[cc.getIdxFromMode(execution, 1, cc.idx)] != 0 {
						cc.idx = cc.codes[cc.getIdxFromMode(execution, 2, cc.idx)]
						break
					}
					cc.idx += 3
				case 6:
					if cc.codes[cc.getIdxFromMode(execution, 1, cc.idx)] == 0 {
						cc.idx = cc.codes[cc.getIdxFromMode(execution, 2, cc.idx)]
						break
					}
					cc.idx += 3
				case 7:
					var toAssign int
					if cc.codes[cc.getIdxFromMode(execution, 1, cc.idx)] < cc.codes[cc.getIdxFromMode(execution, 2, cc.idx)] {
						toAssign = 1
					}
					cc.codes[cc.getIdxFromMode(execution, 3, cc.idx)] = toAssign
					cc.idx += 4
				case 8:
					var toAssign int
					if cc.codes[cc.getIdxFromMode(execution, 1, cc.idx)] == cc.codes[cc.getIdxFromMode(execution, 2, cc.idx)] {
						toAssign = 1
					}
					cc.codes[cc.getIdxFromMode(execution, 3, cc.idx)] = toAssign
					cc.idx += 4
				case 99:
					done <- struct{}{}
					return
				default:
					panic(fmt.Sprintf("OptCode not known: %d", optCode))
				}
			}
		case <-ctx.Done():
			return
		}
	}
}

func (cc *channelCoder) getIdxFromMode(execution int, parameterPosition int, idx int) int {
	mode := execution / int(math.Pow(10, 1+float64(parameterPosition)))
	mode %= 10
	if mode == 1 {
		return idx + parameterPosition
	}
	return cc.codes[idx+parameterPosition]
}

func ChannelCoderFindMaxThrusterSignal(codes []int, fromTo FromTo) int {
	return findMaxThrusterSignal(codes, channelCoderMaxSignal, fromTo)
}

type signals struct {
	a int
	b int
	c int
	d int
	e int
}

func (s signals) contains(value int) bool {
	if s.a == value || s.b == value || s.c == value || s.d == value || s.e == value {
		return true
	}
	return false
}

type FromTo struct {
	From int
	To   int
}

type thrusterSignal func(codes []int, signals signals) int

func findMaxThrusterSignal(codes []int, findThrusterSignal thrusterSignal, fromTo FromTo) int {
	below := fromTo.From - 1
	inputs := signals{below, below, below, below, below}
	maxSignal := math.MinInt

	for first := fromTo.From; first < fromTo.To; first++ {
		inputs.a = first

		for second := fromTo.From; second < fromTo.To; second++ {
			if inputs.contains(second) {
				continue
			}
			inputs.b = second

			for third := fromTo.From; third < fromTo.To; third++ {
				if inputs.contains(third) {
					continue
				}
				inputs.c = third

				for fourth := fromTo.From; fourth < fromTo.To; fourth++ {
					if inputs.contains(fourth) {
						continue
					}
					inputs.d = fourth

					for fifth := fromTo.From; fifth < fromTo.To; fifth++ {
						if inputs.contains(fifth) {
							continue
						}
						inputs.e = fifth
						last := findThrusterSignal(codes, inputs)
						if last > maxSignal {
							maxSignal = last
						}
						inputs.e = below
					}
					inputs.d = below
				}
				inputs.c = below
			}
			inputs.b = below
		}
		inputs.a = below
	}
	return maxSignal
}

func channelCoderMaxSignal(codes []int, signals signals) int {
	a := newChannelCoder("a", codes, []int{signals.a, 0})
	b := newChannelCoder("b", codes, []int{signals.b})
	c := newChannelCoder("c", codes, []int{signals.c})
	d := newChannelCoder("d", codes, []int{signals.d})
	e := newChannelCoder("e", codes, []int{signals.e})
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()
	_ = a.attach(ctx, &b)
	_ = b.attach(ctx, &c)
	_ = c.attach(ctx, &d)
	_ = d.attach(ctx, &e)
	doneE := e.attach(ctx, &a)

	<-doneE

	return e.getLatestSignal()
}

File: year2019/cs/Day7.Tests/ChannelCoder.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Channels;
using System.Threading.Tasks;

namespace Day7.Tests;

internal class ChannelCoder : IDisposable
{
    public string Identifier { get; }
    public Task? Work { get; private set; }
    private readonly IList<int> codes;
    private readonly Channel<int> channel;
    private readonly CancellationTokenSource cts;
    private int idx;
    private ChannelCoder? observer;
    private readonly List<int> outputs;

    private ChannelCoder(string identifier, IList<int> codes)
    {
        Identifier = identifier;
        this.codes = codes;
        channel = Channel.CreateUnbounded<int>();
        cts = new CancellationTokenSource();
        idx = 0;
        outputs = new List<int>();
    }

    internal static int FindMaxThrusterSignal(IList<int> codes)
    {
        return CoderAlgorithms.FindMaxThrusterSignal(codes, FindThrusterSignal());
    }

    private static Func<IList<int>, (int A, int B, int C, int D, int E), Task<int>> FindThrusterSignal()
    {
        return async (codes, inputs) =>
        {
            var coders = await SetupAmplifiers(codes, inputs)
                .ConfigureAwait(false);
            var coder = coders["E"];

            await coder.Work!
                .ConfigureAwait(false);
            var thrusterSignal = coder.GetLargestSignal();
            return thrusterSignal;
        };
    }

    internal static async Task<IDictionary<string, ChannelCoder>> SetupAmplifiers(IList<int> codes,
        (int A, int B, int C, int D, int E) inputs)
    {
        var a = await CreateChannelCoder("A", new List<int>(codes), new List<int> { inputs.A, 0});
        var b = await CreateChannelCoder("B", new List<int>(codes), new List<int> { inputs.B });
        var c = await CreateChannelCoder("C", new List<int>(codes), new List<int> { inputs.C });
        var d = await CreateChannelCoder("D", new List<int>(codes), new List<int> { inputs.D });
        var e = await CreateChannelCoder("E", new List<int>(codes), new List<int> { inputs.E });
        a.Attach(b);
        b.Attach(c);
        c.Attach(d);
        d.Attach(e);
        e.Attach(a);
        return new Dictionary<string, ChannelCoder>()
        {
            {a.Identifier, a},
            {b.Identifier, b},
            {c.Identifier, c},
            {d.Identifier, d},
            {e.Identifier, e},
        };
    }

    internal static async Task<ChannelCoder> CreateChannelCoder(string identifier, IList<int> codes, IList<int> initialInputs)
    {
        var coder = new ChannelCoder(identifier, codes);
        foreach (var initialInput in initialInputs)
        {
            await coder.channel.Writer.WriteAsync( initialInput, coder.cts.Token);
        }
        return coder;
    }

    internal int GetLargestSignal()
    {
        return outputs[^1];
    }


    internal async Task Notify(int input, CancellationToken ct)
    {
        await channel.Writer.WriteAsync(input, ct);
    }

    internal void Attach(ChannelCoder coder)
    {
        observer = coder;
        Work = Task.Run(Run);
    }

    private async Task Run()
    {
        await foreach (var input in channel.Reader.ReadAllAsync(cts.Token))
        {
            var inputUsed = false;
            var keepWhileLoopOn = true;
            do
            {
                var execution = codes[idx];
                switch (execution % 100)
                {
                    case 1:
                        codes[GetIdxFromMode(codes, execution, 3, idx)] =
                            codes[GetIdxFromMode(codes, execution, 2, idx)] + codes[GetIdxFromMode(codes, execution, 1, idx)];
                        idx += 4;
                        break;
                    case 2:
                        codes[GetIdxFromMode(codes, execution, 3, idx)] =
                            codes[GetIdxFromMode(codes, execution, 2, idx)] * codes[GetIdxFromMode(codes, execution, 1, idx)];
                        idx += 4;
                        break;
                    case 3:
                        if (inputUsed)
                        {
                            keepWhileLoopOn = false;
                            break;
                        }
                        codes[GetIdxFromMode(codes, execution, 1, idx)] = input;
                        idx += 2;
                        inputUsed = true;
                        break;
                    case 4:
                        var output = codes[GetIdxFromMode(codes, execution, 1, idx)];
                        await (observer != null ? observer.Notify(output, cts.Token) : null);
                        outputs.Add(output);
                        idx += 2;
                        break;
                    case 5:
                        if (codes[GetIdxFromMode(codes, execution, 1, idx)] != 0)
                        {
                            idx = codes[GetIdxFromMode(codes, execution, 2, idx)];
                            break;
                        }
                        idx += 3;
                        break;
                    case 6:
                        if (codes[GetIdxFromMode(codes, execution, 1, idx)] == 0)
                        {
                            idx = codes[GetIdxFromMode(codes, execution, 2, idx)];
                            break;
                        }
                        idx += 3;
                        break;
                    case 7:
                        codes[GetIdxFromMode(codes, execution, 3, idx)] =
                            codes[GetIdxFromMode(codes, execution, 1, idx)] < codes[GetIdxFromMode(codes, execution, 2, idx)] ?
                                1 : 0;
                        idx += 4;
                        break;
                    case 8:
                        codes[GetIdxFromMode(codes, execution, 3, idx)] =
                            codes[GetIdxFromMode(codes, execution, 1, idx)] == codes[GetIdxFromMode(codes, execution, 2, idx)] ?
                                1 : 0;
                        idx += 4;
                        break;
                    case 99:
                        return;
                    default:
                        throw new Exception($"OptCode not known {execution}");
                }
            } while (keepWhileLoopOn);
        }
    }

    private static int GetIdxFromMode(IList<int> codes, int execution, int parameterPosition, int idx)
    {
        var mode = execution / (int)Math.Pow(10, 1 + parameterPosition);
        mode %= 10;
        return mode == 1 ? idx + parameterPosition : codes[idx + parameterPosition];

    }

    public void Dispose()
    {
        cts.Dispose();
    }
}